### PR TITLE
feat(#60): dashboards with CHT api express metrics

### PR DESCRIPTION
**Pull request by kennsippell**
#60 

### Left Column
Screenshot from 2023-07-18 01-51-33
Dashboard starts with an [replication apdex score](https://en.wikipedia.org/wiki/Apdex). Waiting on some UX research to set these values, but tentatively I've set the following thresholds.

User State | Threshold
-- | --
Satisfied | <90 secs
Tolerating | 90 - 180 secs
Frustrated | > 180 secs

Interesting constraint is that the threshold we monitor for needs to match one of [these hardcoded buckets in CHT 4.3 API](https://github.com/medic/cht-core/pull/8354/files#diff-ca6928872d8d920a49027473c56cc03e5da0d2d94519bb7a69ccf32d51436087R119). So this does not seem particularly agile or easy to change on the fly or per-partner.

Error rate shows % of `get-ids` requests resulting in status code 400-599.

### Right Column

Screenshot from 2023-07-18 01-51-47

Total number of successful and failing replications. 
Rate of replications per second by endpoint (optional. considering removing this)
Replication latency is 50th percentile, 90th percentile, and max replication times

### CODE REVIEW

**Comment by kennsippell** [comment]
2nd dashboard is a techy on based on the [prometheus-api-metrics shared dashboard](https://grafana.com/grafana/dashboards/12230-node-js-dashboard/) with all widgets and settings related to apdex removed.  I don't know what the apdex thresholds should be for general CHT traffic.

File: grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json
Comment (jkuester) [inline grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json:1]: Might be missing something here but where does the `partnerships` in the file name come from?  Is it something we should also include in the title of the dashboard?

**Review by jkuester — COMMENTED** [review]
What is the rational for adding these as brand new dashboards instead of just including them as rows on the `CHT Admin Details` dashboard?  I ask this mostly as a general question in that I am not sure exactly how we should determine "what belongs on a dashboard". I just know from a users perspective, if I want to find data about how long it is taking the server to respond to requests, it is not immediately clear to me if I would find that data on CHT Admin Details, CHT Replication, or CHT API Dashboard.  (c.c. @m5r I would be glad for your feedback on this question as well!)

On one hand it is nice to have all of the panels dependent on a particular scrape config or CHT version all grouped on one dashboard since that makes it easier to document the requirements of a particular dashboard. However, it feels like dependencies (particularly on CHT versions) are bound to evolve over time and things will get more complicated.

My current thinking is that our dashboard design should prioritize the best experience for folks running the latest CHT version (while remaining aware of how changes will impact folks running older versions). Basically, we should put the panels where it makes the most sense to have them (regardless of what CHT version they are dependent on).  If folks with older CHT versions see blank panels that is OK (as long as we indicate in the panel description the minimum CHT version needed for the panel).  

That being said, for this particular case I have not fully made up my mind on where the best place is for the panels to go. :thinking: Seems like if we keep these panels in their own dashboards (like you have them now), then we should plan to break the CHT Admin Details up into separate dashboards (Couch data, Outbound messaging data, etc). This could be done as needed in the future. Open to other ideas though!

File: grafana/provisioning/dashboards/CHT/cht_admin_api_express.json
Comment (jkuester) [inline grafana/provisioning/dashboards/CHT/cht_admin_api_express.json:24]: Minor, but would be nice to include some dashboard links on these new dashboards similar to what the existing ones have (these just make it easier to quickly swap between dashboards).

```suggestion
    "links": [
    {
      "asDropdown": true,
      "icon": "external link",
      "includeVars": true,
      "keepTime": true,
      "tags": [],
      "targetBlank": false,
      "title": "CHT API Server",
      "tooltip": "",
      "type": "dashboards",
      "url": ""
    }
  ],
```

File: grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json
Comment (kennsippell) [reply grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json:1]: I'm hoping these dashboards will be used not by CHT admins, but to help manage partnership relationships.  Check out this [recording](https://us02web.zoom.us/rec/share/6J76ZueDML4YrDnxJAJyCx92wL2issy4ukvhwyNz3hSrYrOhR3saC2ww5uvabJ9L.vFAc1Dvrt2Sy56-m) of topic at ~47minutes.  Passcode is [here](https://medic.slack.com/archives/C0541ANF17U/p1689113619708309).  Here is a more comprehensive overview about [CHT Service Level Objectives with partners](https://docs.google.com/document/d/11BWXPku5siAZpkI3YJnDPkgqInEWhK2XtkccfiVDxzE/edit#). I'm happy to include something to this end in the dashboard title.

**Comment by kennsippell** [comment]
Rationale for adding these into thier own dashboards

`cht_partnerships_replication` I'm hoping will be used by partnerships-level people and not CHT admins. See [this](https://github.com/medic/cht-watchdog/pull/75#discussion_r1270842790). 

`cht_admin_api_express` just has "a lot of stuff". It may be easier to maintain if/when [this changes](https://grafana.com/grafana/dashboards/12230-node-js-dashboard/). It is a lot of deep nerdy performance metrics. I'm unclear when people would want to see them, but I know I want to see them. I suspect are mostly only when investigating performance problems. Maybe this is for more of a core dev user type and not a CHT admin?

When I've worked in monitored environments previously, there were many thousands of dashboards. I didn't know what they were all for, but I could come in and get things done.  I personally preferred having dashboards which were tailored to specific scenarios, rather than a few dashboards which tried to do everything. 

I'd be quite open to including endpoint performance on the existing "CHT details" dashboards - that feels like an interesting level of detail to me for CHT admins.  Thoughts?

File: grafana/provisioning/dashboards/CHT/cht_admin_api_express.json
Comment (kennsippell) [inline grafana/provisioning/dashboards/CHT/cht_admin_api_express.json:2]: Would CHT admins care about this? I suspect this dashboard targets core devs more than CHT admins. It has some deep nerdy metrics. I'm unclear when people would want to see them, but I know I want to see this. I suspect this is mostly useful when looking at performance problems and doing core dev work (?)

It may be easier to maintain if/when [this changes](https://grafana.com/grafana/dashboards/12230-node-js-dashboard/). 

If you agree, I'll update the name to `cht_coredev_api_express` or alike?

**Comment by m5r** [comment]
> What is the rational for adding these as brand new dashboards

> I personally preferred having dashboards which were tailored to specific scenarios/tasks

@kennsippell's take makes sense. I'm not strongly opinionated on either side of the question.

- Going all in with a single mega-dashboard will probably result in meh performances because of having many charts displayed at once.
- Splitting dashboards by persona (i.e. a dashboard tailored for app devs, another one for SREs, another one for CHT admins and so on...) would be nice to give a personalized view of what's happening with a CHT instance but there is bound to be some duplication between dashboards. I don't know if grafana configs can share reusable dashboard "components" but that could be a solution.
- And finally, splitting dashboards by scenario would categorize dashboards neatly with nearly 0 duplicate dashboard config but a single person might need to open many dashboards side by side to get all the information they need.

We can't go wrong with either the first or the third solution, either way we can split dashboards or regroup them without too much overhead

File: grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json
Comment (jkuester) [reply grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json:1]: Got it! Yes! The presentation was very helpful context here!

File: grafana/provisioning/dashboards/CHT/cht_admin_api_express.json
Comment (jkuester) [reply grafana/provisioning/dashboards/CHT/cht_admin_api_express.json:2]: This makes total sense to me!  Super valuable information for development or really deep troubleshooting/debugging, but not useful for day-to-day Admin monitoring.  I agree with keeping this as a separate dashboard and going with a name change to not include `admin`.

**Comment by jkuester** [comment]
Thanks for all the great conversation here @m5r and @kennsippell!  Maybe I was just behind the curve here, but I feel like thing are a lot more clear in my head now regarding organizational strategies for these dashboards!  

It seems to me that if we try to keep dashboards focused on `specific scenarios/tasks` (Mokhtar's `#3`) we would still get pretty much all  the benefits from Mokhtar's `#2` case of splitting ` by persona` (since a given persona would be focused on one or more scenarios/tasks).  So, this seems like a good guiding design principal for us to use!  

And, with that being said, given the extra context @kennsippell provided, I think it makes sense to keep these two as separate dashboards!

**Comment by mrjones-plip** [comment]
Deferring to the feedback from @jkuester  so removing myself as a reviewer.  Lemme know if you want me to jump back in!

File: grafana/provisioning/dashboards/CHT/cht_admin_api_express.json
Comment (kennsippell) [reply grafana/provisioning/dashboards/CHT/cht_admin_api_express.json:2734]: love it

File: grafana/provisioning/dashboards/CHT/cht_admin_api_express.json
Comment (kennsippell) [reply grafana/provisioning/dashboards/CHT/cht_admin_api_express.json:2615]: I removed `restarts_interval` and just set it to 7d permanently. I don't really think it is that valuable. 
I added a description to `interval`.

**Comment by kennsippell** [comment]
> @kennsippell should these dashboards be getting populated with data from the fake-cht?

The API dashboards are expected to work. But the replication dashboards would be empty with fake-cht.  Not clear to me how to get replication dashboards working with fake-cht. I could add the endpoint, but what would ping the endpoint to generate the data? I guess it could be prom (?) If you have ideas I can pursue; but I tested with live CHT.

**Comment by mrjones-plip** [comment]
@jkuester  - with [CHT Core 4.3.0 released](https://docs.communityhealthtoolkit.org/core/releases/4.3.0/) which includes the [API express metrics](https://github.com/medic/cht-core/issues/8426), it'd be good to move this PR along so we can release a matching version Watchdog.  Put this in yer queue when ya get a sec! 

thanks

**Comment by mrjones-plip** [comment]
Ah - I see @kennsippell  is out on holiday until Sept 5th, in case any next steps are blocked until his return!

File: grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json
Comment (jkuester) [inline grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json:1]: It would be nice to have the fake server push data to this dashboard here. It looks like the main thing that is missing is any data for the `/get-ids` endpoint.  I wonder if we could just mock that up in the fake-cht `index.js`?  This seemed to work for me, but open to other ideas:

Update the `prometheusMiddleware` options to include (so that it [matches production](https://github.com/medic/cht-core/blob/55cdaed503e7835a4025755ab84f892a152daeb8/api/src/routing.js#L118)):

```js
  durationBuckets: [
    0.004, 0.007, 0.013, 0.025, 0.05,
    0.1, 0.25, 0.5, 1, 2,
    3, 5, 7.5, 10, 25,
    45, 90, 180, 360, 600,
    1200, 1800, 3600
  ],
```

Listen on new endpoint:

```js
app.get('/get-ids', (req, res) => {
  const timeout = getRandomInt(0, 5000);
  setTimeout(() => {
    if (timeout > 4000) {
      res
        .status(504)
        .send({ message: 'Timeout' });
      return;
    }
    res.json({ id: new Date().getTime() });
  }, timeout);
});
```

Add this to the `app.get('/api/v2/monitoring'` function so that we periodically call `get-ids`:
```js
  fetch('http://fake-cht:8081/get-ids');
```

File: grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json
Comment (jkuester) [inline grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json:141]:
```suggestion
        "show": false,
```

Minor, but since there is only one data series on this panel (for `/get-ids`), we do not really need the legend on the side....

**Review by jkuester — CHANGES_REQUESTED** [review]
:+1: Super excited to get these dashboards in!

File: grafana/provisioning/dashboards/CHT/cht_coredev_api_express.json
Comment (jkuester) [reply grafana/provisioning/dashboards/CHT/cht_coredev_api_express.json:23]: FYI, this is just for adding the navigation links for jumping between dashboards.  Like this one:

image

File: grafana/provisioning/dashboards/CHT/cht_coredev_api_express.json
Comment (kennsippell) [reply grafana/provisioning/dashboards/CHT/cht_coredev_api_express.json:874]: Good catch. I've opened https://github.com/PayU/prometheus-api-metrics/issues/116.  I looked through the code and I understand why this isn't happening. I'll maybe fix + for now, I have adjusted dashboards to work around.

File: grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json
Comment (kennsippell) [reply grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json:1]: > fetch('http://fake-cht:8081/get-ids');

This is what I was missing.

File: grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json
Comment (kennsippell) [reply grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json:154]: I've intended to see two panels - one for /initial-replications/get-ids and one for /replication/get-ids. I'll give this another test, but can you share a screenshot showing what you're hoping to change?

**Comment by kennsippell** [comment]
Thanks for the review @jkuester! 

Good times with fake-cht server now:
image

File: grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json
Comment (jkuester) [reply grafana/provisioning/dashboards/CHT/cht_partnerships_replication.json:154]: Ah, I think when I was looking at this, I though there would only be one time-series displayed. Now with your latest fake code, I can see two and so it makes sense to have the legend.

**Review by jkuester — APPROVED** [review]
LGTM!

**Pull request merged**
